### PR TITLE
[WIP] New Feature: Add database functionality

### PR DESCRIPTION
**Pull request by WassCodeur**
#### Description
This PR adds database functionality to manage data and its menus. The main changes are as follows:

- Added a `restaurant_menus` collection to store information about restaurants and their menus.
- Added verification to avoid duplicates before inserting documents.
- Use of the `pymongo` module to interact with MongoDB.

#### Tests carried out
- Verification of document insertion without duplication.
- Tested basic CRUD operations on the `restaurant_menus` collection.

#### How to test
1. Install dependencies with `pip install -r requirements.txt`.
2. Configure  [MongoDB](https://www.prisma.io/dataguide/mongodb/setting-up-a-local-mongodb-database) and launch the server.
3. Create the `.env` file, copy the contents of `.env.example` and update the database variables with the values from your mongodb database: 
4. Run the `data/db.py` script to test functionality.

#### Additional comments
We welcome any suggestions or improvements!

### CODE REVIEW

**Review by koviubi56 — CHANGES_REQUESTED** [review]
1. Setup instructions for the db should be included in the README.
2. Better docstrings for classes and functions. (see [here](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html))
3. Follow PEP8 by using snake_case for function names.
4. Don't return strings as a result; raise an exception for errors.
5. Simplify if-statements:
  ```py
  # instead of this:
  if self.connected:
     # ...code...
  else:
     raise ...

  # do this:
  if not self.connected:
      raise ...
  # ...code...
  ```
6. `dict.get(x)` might return None. Either set a default or use `dict[x]`
7. Try to use general type hints for parameters:
  ```py
  # x doesn't need to be a list, it just needs to be iterable
  def f(x: Iterable[str]) -> None:
      for something in x:
          g(something)
  ```
8. Consider raising an exception instead of returning None.
9. I'm not sure if we're gonna have tests.

**Comment by WassCodeur** [comment]
Hi @koviubi56 !

Thank you for your feedback

**Review by koviubi56 — APPROVED** [review]
It LGTM, but I'm wondering: is it worth it to use MongoDB? You have to install it, set it up; while there is [`sqlite3`](https://docs.python.org/3/library/sqlite3.html) in stdlib that doesn't need to be set up. So is there a specific reason you chose MongoDB?

**Comment by WassCodeur** [comment]
Hello @koviubi56,

thanks for your comment. In fact, I used mongoDB because of its ability to support several types of data (dynamic schema and rich queries) for one entity, in addition to its scalability and high read/write performance. By the way, you can also use the online version of mongoDB, so there's no need to install the server locally.

**Comment by Vinyzu** [comment]
This is marked as a WIP (Work in Progress).
The current goal of the database is a MongoDB which we can push/pull the savestates to.
We might extend the databases use case if we feel like it fits us well.

File: requirements.txt [hunk 1]
```diff
@@ -3,6 +3,4 @@
 disnake~=2.9.2
 python-dotenv~=1.0.1
 dacite~=1.8.1
-torch~=2.3.1
-sentence-transformers~=3.0.1
-numpy<2
```
Comment (koviubi56) [inline requirements.txt]: Why did these get removed?

File: src/sincere_singularities/data/db.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            collection (str): Collection name
+            data (dict[str, Any]): Data to delete
+
+        Returns:
+            None
+        """
+        if not self.connected:
+            raise ConnectError("Not connected to the database")
+        if self.show_one(collection, data):
+            self.db[collection].delete_one(data)
```
Comment (koviubi56) [inline src/sincere_singularities/data/db.py]: Maybe raise an exception if it didn't exist in the first place? (also in `delete_many`)

File: src/sincere_singularities/data/db.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    def show_all(self, collection: str) -> Iterable[dict[str, Any]]:
```
Comment (koviubi56) [inline src/sincere_singularities/data/db.py]: You can specify that the return value is a list.

File: src/sincere_singularities/data/db.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        Returns:
+            dict[str, Any]: Element found
+        """
+        if not self.connected:
+            raise ConnectError("Not connected to the database")
+
+        element = self.db[collection].find_one(data)
+        if element:
+            return dict(element)
+        return {}
```
Comment (koviubi56) [inline src/sincere_singularities/data/db.py]: Raise instead.

File: src/sincere_singularities/data/db.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+        Returns:
+            None
+        """
+        if not self.connected:
+            raise ConnectError("Not connected to the database")
+
+        element = self.show_one(collection, data)
+
+        if element:
```
Comment (koviubi56) [inline src/sincere_singularities/data/db.py]: Raise exception if it doesn't exist in the first place. (also in update_many)

File: src/sincere_singularities/data/restaurants.json [hunk 6]
```diff
@@ -1,4 +1,52 @@
 [
```
Comment (koviubi56) [inline src/sincere_singularities/data/restaurants.json]: Add points to the restaurants. The first one (index 0) should need 0 points, and the following ones should have more and more (see Pizzaria and Sushi Restaurant for example).

File: src/sincere_singularities/data/savestates.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+class SaveStates:
+    """Save states"""
+
+    def __init__(self) -> None:
+        self.client = DbClient()
+        self.collection = self.client.db.states.name
+        if not self.client.is_connected():
+            raise ConnectError("Not connected to the database")
+
+    def add_user_state(self, data: dict[str, Any]) -> None:
```
Comment (koviubi56) [inline src/sincere_singularities/data/savestates.py]: Maybe you could make this a TypedDict

File: src/sincere_singularities/data/states.json [hunk 8]
```diff
@@ -0,0 +1,63 @@
+[
```
Comment (koviubi56) [inline src/sincere_singularities/data/states.json]: Thanks, but I don't think we need example states.

**Review by koviubi56 — CHANGES_REQUESTED** [review]
Thank you. If you could fix my recommendations, that'd be great. Also, could you implement what I temporarily implemented in `points.py` (between the two comments)?

File: src/sincere_singularities/data/restaurants.json (hunk 6 above)
Comment (koviubi56) [reply src/sincere_singularities/data/restaurants.json]: Actually, we already worked on this, so this addition isn't necessary.

File: README.md [hunk 9]
```diff
@@ -29,3 +30,4 @@ the information overload!
    ```shell
    python -m sincere_singularities
     ```
+2.  Test the bot in your discord server.
```
Comment (Vinyzu) [inline README.md:33]: We will rework the ReadMe in the Presentation phase anyways.

**Review by Vinyzu — APPROVED** [review]
LGTM for now.
Will do some refractoring of the code tomorrow but its important that we get this code in the main branch to implement it tomorrow.

**Review by koviubi56 — APPROVED** [review]
LGTM

**Pull request merged**